### ISSUE TITLE
File Upload not working anymore

### ISSUE THREAD

**Issue by oexler** [terminal42/contao-notification_center#250]
Hello,

I am using contao 4.9.23 and I have problems with the latest version (1.6.4) of the Notification Center.
I guess due to the changes to the temp directory, the files can't be moved.
Probably because the file is not saved by contao 'uploaded' is not set.

In getFileUploadPathForToken , therefore 
[$filePath = sys_get_temp_dir() . '/' . $file['name'];](https://github.com/terminal42/contao-notification_center/blob/d2070fc8b37f956557755077292e0b04b0d00721/library/NotificationCenter/Util/Form.php#L40)
is set.
In my case this is /var/folders/9m/m_sgcx916xjbd5pbscthgphc0000gq/T/file.ext
In \Files::getInstance()->move_uploaded_file, however, $this->strRootDir is prepended to the target, making it /Users/username/Sites/contao//var/folders/9m/m_sgcx916xjbd5pbscthgphc0000gq/T/file.ext, which is probably not a valid path.

**Comment by fritzmg** [comment]
Hm, I see. The changes were introduced in https://github.com/terminal42/contao-notification_center/commit/df188ab03f4ab4d3cd05f37e10b08e69bdd5fe0b
@Toflar it looks like `Contao\Files::move_uploaded_file` needs the path to be relative to project directory.

**Comment by Toflar** [comment]
Should probably just use the native functions then? wdyt?

**Comment by fritzmg** [comment]
Yes, though you might still want to use `Files::validate`.

```php
if (Files::getInstance()->validate($file['tmp_name'], $filePath)) {
    move_uploaded_file($file['tmp_name'], $filePath);
    chmod($filePath, $GLOBALS['TL_CONFIG']['defaultFileChmod']);
}
```

**Comment by fritzmg** [comment]
@oexler can you please test if everything works properly if you replace theses lines

https://github.com/terminal42/contao-notification_center/blob/d2070fc8b37f956557755077292e0b04b0d00721/library/NotificationCenter/Util/Form.php#L41-L42

with 

```php
if (Files::getInstance()->validate($file['tmp_name'], $filePath)) {
    move_uploaded_file($file['tmp_name'], $filePath);
    chmod($filePath, $GLOBALS['TL_CONFIG']['defaultFileChmod']);
}
```

?

**Comment by oexler** [comment]
No this didn't work.
First I got `Class 'NotificationCenter\Util\Files' not found`
Then I added a \ in front of Files but got `Call to protected method Contao\Files::validate() from context 'NotificationCenter\Util\Form'`

**Comment by fritzmg** [comment]
Right. For now you could just use

```php
move_uploaded_file($file['tmp_name'], $filePath);
chmod($filePath, $GLOBALS['TL_CONFIG']['defaultFileChmod']);
```

i.e. without `if (Files::getInstance()->validate($file['tmp_name'], $filePath)) {`.

**Comment by fritzmg** [comment]
@Toflar not sure what we should do though. `Contao\Files` generally only works with the Contao project directory and not outside of it.

**Comment by Toflar** [comment]
Imho we should remove all calls to `File` and `Files` because they make no sense. Uploading to a temporary directory must be possible outside of the `kernel.project_dir`. The `validate()` call can be replaced by `!Validator::isInsecurePath()`.

**Comment by oexler** [comment]
Wouldn't it be better to leave the file in its location and just adjust the filename when sending the email?
Theoretically it could happen that within a short time a file with the same name is uploaded and overwritten before it is sent.

**Comment by Toflar** [comment]
The whole codebase of the NC has really become super old and it's hard to maintain it. We really need a version 2 at some point. I've tried to fix the issue in https://github.com/terminal42/contao-notification_center/commit/e157bf815a0fd39f25d4be228dc95ec89feed950, can you try if this fixes the problem for you?

**Comment by fritzmg** [comment]
@oexler still waiting on feedback - please install `dev-bugfix/file-handling` and test.

**Comment by qzminski** [comment]
Fixed in #252 and 1.6.5.

**Comment by fritzmg** [comment]
@qzminski your verison introduces a new problem on Windows though. Will you create a separate PR for that?

**Comment by qzminski** [comment]
We don't use Windows, so we cannot really test that 🤷‍♂️  Would you be that kind and see if your proposal with https://packagist.org/packages/webmozart/path-util works on Windows?

**Comment by oexler** [comment]
Sorry for the delay on my feedback. The current version 1.6.5 seems to work for me.

**Comment by colonelpanick** [comment]
Hi all, am I the only one still seeing this problem with NC 1.6.10 (Contao 4.9.22, PHP 7.4.28, Linux hosting)? 

File upload via frontend form does not result in an email attachment. Interestingly, the correct file name is output if token ##form_[uploadfield]## is also used in the message body ("/tmp/[filen.ame]"). No special characters or other shenanigans used in file name.

Attaching the file to a message fails silently, no errors or log-entries.
